R/estimate_uncertainty.R

```diff
@@ -135,73 +140,211 @@ estimate_uncertainty <- function(
+    cli_warn(
+      "Only the first {n_iters} retrospective nowcast times were used."
+    )
+  }
+
+  agg <- as.matrix(ref_time_aggregator(list_of_obs[[1]]))
+  ncol_agg <- ncol(agg)
+
+  if (ncol_agg != ncol(list_of_obs[[1]])) {
+    cli_abort(message = c(
+      "`ref_time_aggregator` must return a matrix with ",
+      "the same number of columns as the input observations"
+    ))
+  }
+
+
+
   # Each row is retrospective nowcast date, each column is a horizon (i.e
   # columns are not delays, but horizons, and each cell contains a total
   # value corresponding to that horizon -- the total expected value to add
   exp_to_add <-
-    to_add_already_observed <- matrix(NA, nrow = n, ncol = n_possible_horizons)
-  for (i in seq_len(n_iters)) { # Seq along retrospective forecast dates
-    # Rretrospective nowcast as of i delays ago
+    to_add_already_observed <- matrix(NA,
+      nrow = n_iters,
+      ncol = n_possible_horizons
+    )
+
+
+
+  for (i in seq_len(n_iters)) {
+    # For each individual retrospective nowcast, extract matrices we need from
+    # the corresponding elements in the list.
     nowcast_i <- list_of_ncs[[i]]
-    # Remove the last i observations
     trunc_matr_observed <- list_of_obs[[i]]
     triangle_observed <- list_of_rts[[i]]
-    max_t <- nrow(trunc_matr_observed)
-    n_horizons <- min(max_t - k + 1, n_possible_horizons)
-    if (i == 1 && n_horizons < n_possible_horizons) {
-      cli_abort(
-        message = c(
-          sprintf(
-            "Requested window size k=%i is too large to generate sufficient nowcasts for the required forecast horizons for all available nowcast matrices.", # nolint
-            k
-          )
-        )
-      )
-    }
-    if (n_horizons < n_possible_horizons) {
-      cli_warn(message = c(
-        sprintf(
-          "Requested window size k=%i is too large to generate nowcasts for all forecast horizons in matrix %i. ", # nolint
-          k,
-          i
-        )
-      ))
-    }
-    # Take the reporting triangle and look at one row at a time, which
-    # corresponds to one horizon
-    for (d in 1:n_horizons) {
-      start_row <- max_t - d - k + 2
-      end_row <- max_t - d + 1
-      obs <- trunc_matr_observed[start_row:end_row, ]
-      nowcast <- nowcast_i[start_row:end_row, ]
-      indices_nowcast <- is.na(triangle_observed[start_row:end_row, ])
-      indices_observed <- !is.na(obs)
-      # Function to aggregate is always applied after the matrix has been
-      # summed across delays
-      exp_to_add[i, d] <- fun_to_aggregate(
-        rowSums(as.matrix(nowcast *
-          indices_nowcast * indices_observed), na.rm = TRUE),
-        na.rm = TRUE
-      )
-      to_add_already_observed[i, d] <- fun_to_aggregate(
-        rowSums(as.matrix(
-          obs * indices_nowcast * indices_observed
-        ), na.rm = TRUE),
-        na.rm = TRUE
+
+    # Apply the aggregation to the truncated observations, the nowcast,
+    # and the reporting triangle.
```

Comment:
see later comment on if this is the right name

Reply:
I think my preference is to keep ref time aggregator and delay aggregator. I changed the argument to `estimate_uncertainty` to `uncertainty_model` as it is about how you are estimating the uncertainty from the matrices rather than what the specific type of observation model it is that you are using. 

`fit_by_horizon()` then takes in the `observation_model` (which is `fit_nb`).

Reply:
why?

Reply:
Why what? Why keep ref and delay? Because thats easier to remember then knowing ref = row and delay = column.

Why `uncertainty_model`? Because it is a description of how we are computing uncertainty not the observation model we are assuming between the predictions and observations -- maybe should be `uncertainty_operation` or `uncertainty_function`?